contracts/rewards/src/contract/execute.rs

```diff
@@ -88,16 +89,45 @@ where
         Ok(())
     }
 
-    pub fn distribute_rewards(
+    pub fn process_rewards(
         &mut self,
-        _contract: Addr,
-        _block_height: u64,
-        _count: Option<u64>,
+        contract: Addr,
+        block_height: u64,
+        count: Option<u64>,
     ) -> Result<HashMap<Addr, Uint256>, ContractError> {
-        todo!()
+        let count = count.unwrap_or(DEFAULT_EPOCHS_TO_PROCESS);
+
+        let cur_epoch = self.current_epoch(block_height)?;
+        let start_epoch = self
+            .store
+            .load_rewards_watermark(contract.clone())?
+            .map_or(0, |last_processed| last_processed + 1); // start at 0 if no stored watermark
+        let mut epoch_to_process = start_epoch;
+
+        // running accumulation of rewards per worker
+        let mut accumulated_rewards = HashMap::new();
+
+        // we only distribute rewards for a given epoch T starting in epoch T + 2
+        while epoch_to_process + 2 <= cur_epoch.epoch_num && epoch_to_process < start_epoch + count
+        {
+            let new_rewards = self.process_rewards_for_epoch(contract.clone(), epoch_to_process)?;
+            accumulated_rewards = fold_rewards(accumulated_rewards, new_rewards);
+
+            epoch_to_process += 1;
+        }
+
+        // update the watermark if we actually processed any epochs
+        if epoch_to_process != start_epoch {
+            self.store
+                .save_rewards_watermark(contract, epoch_to_process - 1)?;
+        } else {
+            return Err(ContractError::NoRewardsToDistribute.into());
+        }
+
+        Ok(accumulated_rewards)
```

Comment:
```suggestion
        let count = count.unwrap_or(DEFAULT_EPOCHS_TO_PROCESS);
        let cur_epoch = self.current_epoch(block_height)?;

        let from = self
            .store
            .load_rewards_watermark(contract.clone())?
            .map_or(0, |last_processed| last_processed + 1);
        let to = std::cmp::min(from + count - 1, cur_epoch.epoch_num - 2);
        let rewards =
            self.process_rewards_for_epochs(contract.clone(), from, to, HashMap::new())?;

        self.store.save_rewards_watermark(contract, to)?;

        Ok(rewards)
    }

    fn process_rewards_for_epochs(
        &mut self,
        contract: Addr,
        from: u64,
        to: u64,
        rewards: HashMap<Addr, Uint256>,
    ) -> Result<HashMap<Addr, Uint256>, ContractError> {
        if from > to {
            return Ok(rewards);
        }

        let new_rewards = self.process_rewards_for_epoch(contract.clone(), from)?;
        self.process_rewards_for_epochs(contract, from + 1, to, fold_rewards(rewards, new_rewards))
    }

    fn process_rewards_for_epoch(
        &mut self,
        contract: Addr,
        epoch_num: u64,
    ) -> Result<HashMap<Addr, Uint256>, ContractError> {
        match self.store.load_epoch_tally(contract, epoch_num)? {
            None => Ok(HashMap::new()), // no rewards if there is no tally
            Some(tally) => self.process_epoch_tally(tally),
        }
    }
```
I can't apply the suggestion to the function `process_rewards_for_epoch`, but the `.clone` there is not necessary at all.